difftest/dpi_t1rocketemu/src/bus/disp.rs

```diff
@@ -0,0 +1,149 @@
+
+impl ShadowDevice for DisplayDevice {
+  fn new() -> Box<dyn ShadowDevice>
+  where
+    Self: Sized,
+  {
+    Box::new(Self {
+      vram: vec![0u8; (DISPLAY_WIDTH * DISPLAY_HEIGHT * 3) as usize],
+      frame_counter: 0,
+    })
+  }
+
+  fn read_mem(&self, addr: usize, size: usize) -> Vec<u8> {
+    let mut result = vec![0u8; size];
+    let start = addr;
+    let end = addr + size;
+
+    let vram_part_end = std::cmp::min(REG_FLUSH, end);
+    let vram_part_size = if vram_part_end > start {
+      vram_part_end - start
+    } else {
+      0
+    };
+
+    let reg_part_start = std::cmp::max(REG_FLUSH, start);
+    let reg_part_size = if end > reg_part_start {
+      end - reg_part_start
+    } else {
+      0
+    };
+
+    assert_eq!(
+      vram_part_size + reg_part_size,
+      size,
+      "Size calculation done wrong?"
+    );
+
+    self.vram_read(&mut result[0..vram_part_size], start, vram_part_size);
+    self.reg_read(
+      &mut result[vram_part_size..size],
+      reg_part_start - REG_FLUSH,
+      reg_part_size,
+    );
+
+    result
+  }
+
+  fn write_mem(&mut self, addr: usize, data: u8) {
+    if addr < self.vram.len() {
+      self.vram[addr] = data;
+    } else if addr == REG_FLUSH {
+      self.encode_png().unwrap();
+      self.frame_counter += 1;
+    }
+  }
+
+  fn write_mem_chunk(&mut self, addr: usize, size: usize, strobe: Option<&[bool]>, data: &[u8]) {
+    let start = addr;
+    let end = addr + size;
+
+    if end <= self.vram.len() {
+      if let Some(masks) = strobe {
+        masks.iter().enumerate().for_each(|(i, mask)| {
+          if *mask {
+            self.vram[addr + i] = data[i];
+          }
+        })
+      } else {
+        let start = addr;
+        let end = addr + size;
+        self.vram[start..end].copy_from_slice(data);
+      }
+    } else {
+      if let Some(masks) = strobe {
+        masks.iter().enumerate().for_each(|(i, mask)| {
+          if *mask {
+            self.write_mem(addr + i, data[i]);
+          }
+        })
+      } else {
+        for i in start..end {
+          self.write_mem(addr + i, data[i]);
+        }
+      }
+    }
+  }
+}
+
+impl DisplayDevice {
+  fn vram_read(&self, buffer: &mut [u8], offset: usize, size: usize) {
+    let start = std::cmp::min(self.vram.len(), offset);
+    let end = std::cmp::min(self.vram.len(), offset + size);
+    buffer[0..(start - end)].copy_from_slice(&self.vram[start..end]);
+  }
+
+  fn reg_read(&self, buffer: &mut [u8], offset: usize, size: usize) {
+    let counter = self.frame_counter.to_le_bytes().into_iter();
+    let width = (DISPLAY_WIDTH as u16).to_le_bytes().into_iter();
+    let height = (DISPLAY_HEIGHT as u16).to_le_bytes().into_iter();
+    let reg_space: Vec<u8> = counter.chain(width).chain(height).collect();
+
+    let start = std::cmp::min(reg_space.len(), offset);
+    let end = std::cmp::min(reg_space.len(), offset + size);
+    buffer[0..(end - start)].copy_from_slice(&reg_space[start..end]);
+  }
```

Comment:
We don't need to support one transaction read multiple MMIO control registers. Instead, we could requires all transactions are non-burst and naturally aligned, otherwise we report an error. A simple 'match' should suffice.

Reply:
While I can assume all read access to be aligned and never cross register boundry, I think the software can read part of one register (for example width and height is packed in one 32-bit register) and dealing with different sizes would not lead to a cleaner implementation than current generic one.